### PR TITLE
Make markdown image dimensions configurable using a custom image renderer

### LINKED ISSUE #14803: Support a way to set image size in MarkedString
**User story**
As an extension developer,
I would like to be able to set image sizes in markdown,
So that I can provide nicer images in the hover widget.

**Tech. background:**
VSCode uses marked as markdown processor. Unfortunately this does not provides a way to set width and height for images out of the box...
There is a way to configure a special renderer for it e.g.:
```javascript
 renderer.image = function(href, title, text) {                                                                                                                                                                       
    if (title) {                                                                                                              
        var size = title.split('x');                                                                                          
        if (size[1]) {                                                                                                        
            size = 'width=' + size[0] + ' height=' + size[1];                                                                 
        } else {                                                                                                              
            size = 'width=' + size[0];                                                                                        
        }                                                                                                                     
    } else {                                                                                                                  
        size = '';                                                                                                            
    }                                                                                                                         
    return ('<img src="' + href + '" alt="' + text + '" ' + size + '>');                                                      
};
```

This would allow to set image size with a syntax:
```
![alt](http://img.png "420x420")
```

Perhaps a better syntax would be something similar to the one [jira](https://jira.atlassian.com/browse/JRA-30476) uses:
```
!projectavatar.png|height=250,width=250!
```
**Based on this a preffered syntax could be:**
```
![alt](http://img.png|"height=420,width=420")
```

**Relevant issues**
https://github.com/chjj/marked/issues/339
https://github.com/chjj/marked/pull/535
**Relevant commits**
There is already a PR about this in the marked repo (however it is pending since 2015 Jan.)
https://github.com/chjj/marked/pull/535/commits/1a6923a7dba39414e849796f2dd3c5541838cfd8

### PR DESCRIPTION
**Pull request by kisstkondoros**
In the newly introduced syntax extension the width and height of an image
can be given in the image title using following format:
`image`.
Both width and height parameters are optional, if they are not given the
width and height attributes remains unset on the image element

This resolves #14803

### CODE REVIEW

File: src/vs/base/browser/htmlContentRenderer.ts [hunk 1]
```diff
@@ -90,6 +90,29 @@ function _renderHtml(content: IHTMLContentElement, options: RenderOptions = {}):
...
 
 		const renderer = new marked.Renderer();
+		renderer.image = (href: string, title: string, text: string) => {
+			let attributes = '';
+			if (title) {
+				const splitted = title.split('|');
+				title = splitted[0];
+				const parameters = splitted[1];
+				if (parameters) {
+					const heightFromParams = new RegExp(/height=(\d+)/).exec(parameters);
```
Comment (mjbvz) [inline src/vs/base/browser/htmlContentRenderer.ts]: [nit] Just use regexp literal, not `new RegExp`

File: src/vs/base/browser/htmlContentRenderer.ts [hunk 2]
```diff
@@ -90,6 +90,29 @@ function _renderHtml(content: IHTMLContentElement, options: RenderOptions = {}):
 		const withInnerHTML = new TPromise(c => signalInnerHTML = c);
 
 		const renderer = new marked.Renderer();
+		renderer.image = (href: string, title: string, text: string) => {
+			let attributes = '';
+			if (title) {
+				const splitted = title.split('|');
+				title = splitted[0];
```
Comment (mjbvz) [inline src/vs/base/browser/htmlContentRenderer.ts]: Looks like the actual title is getting lost now. Add it back into the attributes

File: src/vs/base/browser/htmlContentRenderer.ts [hunk 3]
```diff
@@ -90,6 +90,29 @@ function _renderHtml(content: IHTMLContentElement, options: RenderOptions = {}):
...
+				const parameters = splitted[1];
+				if (parameters) {
+					const heightFromParams = new RegExp(/height=(\d+)/).exec(parameters);
+					const widthFromParams = new RegExp(/width=(\d+)/).exec(parameters);
+					const height = (heightFromParams && heightFromParams[1]);
+					const width = (widthFromParams && widthFromParams[1]);
+					const dimensions = [];
+					if (width) {
```
Comment (mjbvz) [inline src/vs/base/browser/htmlContentRenderer.ts]: Change this check to `isFinite` to correctly handle zero width/heights

**Review by mjbvz — CHANGES_REQUESTED** [review]
Added a few suggestions. I like the idea, but am not sure about overloading the title to support image size, at least without additional checks to make sure we still correctly support titles that contain `|` characters.

Would using the image source instead work?

``` md
![alt](image.png|width=200,height=300 'title')
```

I'm not sure if marked has already encoded the `|` to  a `%7C` when the render function is called or not.

**Comment by kisstkondoros** [comment]
@mjbvz I've changed the implementation according to your suggestions

**Comment by neilsustc** [comment]
@kisstkondoros Thanks a lot!

**Pull request merged**